### PR TITLE
#1070 Search component (1)

### LINKED ISSUE #1070: Programs: Search database
From: https://github.com/openmsupply/open-msupply/issues/644#issue-1383248595

> #### `Mother`:
> 
>   - only when patient is an infant (Poruan checking the age threshold, most likely expressed in months. We are considering creating a specific program for infants)
>  - the objective is to link the patient record to the mother's patient record. Would it be possible to select mother's patient ID from a dropdown ?

This will be a variation on the standard "Autocomplete" JSON Forms component, but with a database lookup for its source. 

It also needs to:
- only show if patient is infant
- collect all relevant "Mother" info (probably IDs and name)
- maybe a hyperlink to the actual Mother record

I have a suggestion for how to modify existing components to allow more dynamic logic (i.e. database lookups), but I'll need to sit down with someone to clarify the requirements and nut out the details. (@clemens-msupply ?)

### PR DESCRIPTION
**Pull request by CarlosNZ**
Fixes #1070 (and provides basis for #1071)

This was getting a bit big, so I've split this into two parts -- ~~another PR will follow.~~ (Part 2: #1230 )

# 👩🏻‍💻 What does this PR do? 

I tried a few different approaches to this (hence took quite a while), and I finally settled on breaking it up into two distinct components -- one for lookups using "user input" as a source (e.g. typing a code to find a patient), and one that uses existing document data as a source.

There's still only one main "Search" component though, we just need to specify the "source" as an option (either "user" or "document" at this point) -- the Search component is then just a wrapper for either of the two child components.

This PR just includes the "User input" component, I'm still working on the "Document input" component.

The search with user input component looks like this initially:

Screenshot 2023-02-22 at 3 17 58 PM

And when searching:
Screenshot 2023-02-22 at 3 18 07 PM

And after selection:
Screenshot 2023-02-22 at 3 18 14 PM

The display strings are fully customisable if we want to override the defaults.

So the options for the "Search" component are:

- `source`: either `user` or `document` (only `user` done so far)

And for the "User" source, (all except `query` are optional):
- `query`: name of the predefined query, as defined in `useSearchQueries.tsx` 
- `optionString`: pattern to define how results show up in the drop-down
- `displayString`: pattern to define how the selected result shows up in the UI
- `saveFields`: an array of field names to save to the document data
- `placeholderText`: "prompt" text for the Input field UI

# 🧪 How has/should this change been tested? 

Tested using newly created patients. Currently, the main implementation is in the "HIV Care and Treatment" program enrolment form. Please try creating patients with different fields filled in and see the resulting outcomes when searched for. Also, if you want you can play round with the above options in the ui schema (`hiv_care_program_ui_schema.json`).

## 💌 Any notes for the reviewer?

See inline comments for further info

### CODE REVIEW

File: client/packages/common/src/utils/formatters/JSXFormatters.tsx
Comment (CarlosNZ) [inline client/packages/common/src/utils/formatters/JSXFormatters.tsx]: This seemed like a useful utility that might come in handy elsewhere, so I've put it in a new utils file. Not sure of the best place for this, but didn't quite seem to fit in with the other utils. Any particular suggestions, @mark-prins ?

File: client/packages/common/src/utils/regex/RegexUtils.tsx
Comment (CarlosNZ) [inline client/packages/common/src/utils/regex/RegexUtils.tsx:64]: Was still returning "null" values as strings, so this uses the fallback instead

File: client/packages/common/src/types/schema.ts
Comment (CarlosNZ) [inline client/packages/common/src/types/schema.ts:2828]: Duplicate definitions -- caused by merge from base/conflict resolution

File: client/packages/common/src/utils/formatters/JSXFormatters.tsx
Comment (CarlosNZ) [inline client/packages/common/src/utils/formatters/JSXFormatters.tsx:17]: This seemed like a useful utility that might come in handy elsewhere, so I've put it in a new utils file. Not sure of the best place for this, but didn't quite seem to fit in with the other utils. Any particular suggestions, @mark-prins ?

File: client/packages/programs/src/JsonForms/components/Search/Search.tsx
Comment (CarlosNZ) [inline client/packages/programs/src/JsonForms/components/Search/Search.tsx:31]: To-do

File: client/packages/programs/src/JsonForms/components/Search/Search.tsx
Comment (CarlosNZ) [inline client/packages/programs/src/JsonForms/components/Search/Search.tsx:33]: Will fix in next PR

File: client/packages/programs/src/JsonForms/components/Search/SearchWithUserSource.tsx
Comment (CarlosNZ) [inline client/packages/programs/src/JsonForms/components/Search/SearchWithUserSource.tsx:44]: Having a separate options validator in the child components seemed like a simpler option than a complex discriminated union at the top level.

File: client/packages/programs/src/JsonForms/components/Search/SearchWithUserSource.tsx
Comment (CarlosNZ) [inline client/packages/programs/src/JsonForms/components/Search/SearchWithUserSource.tsx:73]: This ensures the results get cleared if user input goes below `minChars` after being longer.

File: client/packages/system/src/Patient/api/hooks/document/usePatients.ts
Comment (CarlosNZ) [inline client/packages/system/src/Patient/api/hooks/document/usePatients.ts:12]: Needed to be able to add query params beyond what is in the URL.

File: client/packages/system/src/Patient/api/operations.graphql
Comment (CarlosNZ) [inline client/packages/system/src/Patient/api/operations.graphql:12]: Get the whole data object.

File: server/service/src/sync/program_schemas/hiv_care_program_ui_schema.json
Comment (CarlosNZ) [inline server/service/src/sync/program_schemas/hiv_care_program_ui_schema.json:133]: Ignore -- I'll update this when I do the document-source component.

File: client/packages/system/src/Patient/api/hooks/document/usePatients.ts
Comment (clemens-msupply) [reply client/packages/system/src/Patient/api/hooks/document/usePatients.ts:12]: Related comment, I found the hooks sometimes confusing because its hard to figure out where the hook is getting its params from and thus where the hook can be used. Find it easier to read when the params are passed in explicitly...

File: client/packages/programs/src/JsonForms/components/Search/useSearchQueries.tsx
Comment (clemens-msupply) [inline client/packages/programs/src/JsonForms/components/Search/useSearchQueries.tsx:81]: Could you use `white-space: pre` instead? or would this not work?

File: client/packages/programs/src/JsonForms/components/Search/useSearchQueries.tsx
Comment (clemens-msupply) [inline client/packages/programs/src/JsonForms/components/Search/useSearchQueries.tsx:7]: First think I tried was searching by name :) think we need to support this (search by code OR firstName OR lastName) but this might need another change to the BE api...

File: client/packages/programs/src/JsonForms/components/Search/SearchWithUserSource.tsx
Comment (clemens-msupply) [reply client/packages/programs/src/JsonForms/components/Search/SearchWithUserSource.tsx:73]: maybe put this comment in the code...

**Review by clemens-msupply — CHANGES_REQUESTED** [review]
Looks good thanks! only one request to move a getSdk call into a new hook for consistency.

The search results are displayed as a text. I assumed its planned to put them into the correct mother entries?

**Comment by mark-prins** [comment]
I was confused when using the control - I didn't know what the patient codes were, only that the name was Tina. I tried typing stuff in, and kept getting `No results` Didn't realise that I had to type 3 characters in order to get a result. This will need helper text or a different error message

Screenshot 2023-02-23 at 2 50 33 PM

Screenshot 2023-02-23 at 2 50 40 PM

File: client/packages/common/src/utils/formatters/JSXFormatters.tsx
Comment (mark-prins) [reply client/packages/common/src/utils/formatters/JSXFormatters.tsx:17]: Not required - just add `whitespace` when displaying, like so:
`<Typography style={{ whiteSpace: 'pre' }}>`

Other comments.. the use of `<br/>` tags as layout controls is frowned upon, though I'd use them in this example! Ideally each line should be wrapped in `Typography` 🙄

File: client/packages/programs/src/JsonForms/components/Search/SearchWithUserSource.tsx
Comment (mark-prins) [inline client/packages/programs/src/JsonForms/components/Search/SearchWithUserSource.tsx:159]:
```suggestion
            height="20px"
            width="20px"
```

File: client/packages/programs/src/JsonForms/components/Search/useSearchQueries.tsx
Comment (mark-prins) [inline client/packages/programs/src/JsonForms/components/Search/useSearchQueries.tsx:91]:
```suggestion
          <Typography style={{ whiteSpace: 'pre' }}>
            {displayString
              ? formatTemplateString(displayString, data, '')
              : formatTemplateString(
                  '${firstName} ${lastName} (${code})\n${email}\n${document.data.contactDetails[0].address1}\n${document.data.contactDetails[0].address2}',
                  data,
                  ''
                )}
          </Typography>
```

**Review by mark-prins — CHANGES_REQUESTED** [review]
Very nice!

File: client/packages/common/src/utils/formatters/JSXFormatters.tsx
Comment (CarlosNZ) [reply client/packages/common/src/utils/formatters/JSXFormatters.tsx:17]: Cool, that works thanks. The only thing we lose is I added a little bit of additional functionality (in the next PR #1230) that also strips out blank lines, which is useful if a field in the displayString is not present. Minor thing, but looks a little nicer. I'll have a think about another way to achieve the same thing when I revisit that PR shortly.

File: client/packages/programs/src/JsonForms/components/Search/useSearchQueries.tsx
Comment (CarlosNZ) [reply client/packages/programs/src/JsonForms/components/Search/useSearchQueries.tsx]: Yeah, have done that: c2d76b82d

Loses one little bit of functionality as per my response to Mark above: https://github.com/openmsupply/open-msupply/pull/1223#discussion_r1119321186

Will address in next PR

File: client/packages/programs/src/JsonForms/components/Search/SearchWithUserSource.tsx
Comment (CarlosNZ) [reply client/packages/programs/src/JsonForms/components/Search/SearchWithUserSource.tsx:74]: 👍🏼  2d4328000

File: client/packages/programs/src/JsonForms/components/Search/useSearchQueries.tsx
Comment (CarlosNZ) [reply client/packages/programs/src/JsonForms/components/Search/useSearchQueries.tsx:7]: Yeah, the issue (#1070) specifically asks for searching by patient code. Can other searches, but this would be a new issue I think (when it's required).

File: client/packages/system/src/Patient/api/hooks/document/usePatients.ts
Comment (CarlosNZ) [reply client/packages/system/src/Patient/api/hooks/document/usePatients.ts:12]: Not sure what you're referring to. You mean in this case, or for the database hooks in general?

**Comment by CarlosNZ** [comment]
> Looks good thanks! only one request to move a getSdk call into a new hook for consistency.

Not seeing this comment/request anywhere else? Am I missing it? Or could you elaborate?

> The search results are displayed as a text. I assumed its planned to put them into the correct mother entries?

The selected result gets saved in the JSON document at the specified "scope", as expected, which in this case is the `#/properties/mother` field. What else did you mean by "correct mother entries"?

File: client/packages/system/src/Patient/api/hooks/document/usePatients.ts
Comment (clemens-msupply) [reply client/packages/system/src/Patient/api/hooks/document/usePatients.ts:12]: No was a general comment for hooks where it might be simpler to pass in the params rather than implicitly fetch them from the url...

**Comment by clemens-msupply** [comment]
> > Looks good thanks! only one request to move a getSdk call into a new hook for consistency.
> 
> Not seeing this comment/request anywhere else? Am I missing it? Or could you elaborate?

sorry think, I didn't submitted my previous comment, I added a new one...

> > The search results are displayed as a text. I assumed its planned to put them into the correct mother entries?
> 
> The selected result gets saved in the JSON document at the specified "scope", as expected, which in this case is the `#/properties/mother` field. What else did you mean by "correct mother entries"?

ok cool thanks, wasn't sure about it

**Comment by CarlosNZ** [comment]
> I was confused when using the control - I didn't know what the patient codes were, only that the name was Tina. I tried typing stuff in, and kept getting `No results` Didn't realise that I had to type 3 characters in order to get a result. This will need helper text or a different error message
> 
> Screenshot 2023-02-23 at 2 50 33 PM
> 
> Screenshot 2023-02-23 at 2 50 40 PM

Okay, have added more informative text in the "No results" area when below the minimum:

Screenshot 2023-02-28 at 12 31 22 PM

6a96cd20e

Had to add an additional state variable to keep track of what the "No results" text should be -- otherwise it would display "No results" as soon as you hit 3 characters, and then change to a result once the debounce and search finished. So now it includes the debounce time as part of the "search" in order to avoid this misleading comment. (Open to other suggestions how to achieve this if you have any?)

File: client/packages/programs/src/JsonForms/components/Search/useSearchQueries.tsx
Comment (CarlosNZ) [reply client/packages/programs/src/JsonForms/components/Search/useSearchQueries.tsx]: Yeah okay, I think this works: 17c10c4a6

```ts
const patientQueries = usePatientApi();
```

I was conscious of not *conditionally* calling a hook (within the SearchQuery map), cos React doesn't let you do that, and I didn't want to have database lookups firing off if they're not being used, so I I couldn't use the next level up (in `usePatient`). 

But yeah, I find the 50 million layers of abstraction we have in place quite confusing. It's always a nightmare trying to make sense of it. Would love someone to explain the logic of it to me sometime.

Thanks.

File: client/packages/programs/src/JsonForms/components/Search/SearchWithUserSource.tsx
Comment (CarlosNZ) [reply client/packages/programs/src/JsonForms/components/Search/SearchWithUserSource.tsx:167]: Yeah, it's annoying how VSCode auto-prompts to that -- I normally catch them and fix them :) 

d4910a923

File: client/packages/programs/src/JsonForms/components/Search/useSearchQueries.tsx
Comment (clemens-msupply) [reply client/packages/programs/src/JsonForms/components/Search/useSearchQueries.tsx]: > But yeah, I find the 50 million layers of abstraction we have in place quite confusing. It's always a nightmare trying to make sense of it. Would love someone to explain the logic of it to me sometime.

I agree I also find it harder to read and write than necessary. @mark-prins maybe we can discuss this at some point?

**Review by clemens-msupply — CHANGES_REQUESTED** [review]
Found one bug: you can select yourself as mother :)

File: client/packages/programs/src/JsonForms/components/Search/useSearchQueries.tsx
Comment (clemens-msupply) [reply client/packages/programs/src/JsonForms/components/Search/useSearchQueries.tsx:6]: Ok cool, lets see what @richardmoizeau thinks. After having another look at the current implementation I have the feeling we need to support manual mother entries as well...

**Comment by CarlosNZ** [comment]
> Found one bug: you can select yourself as mother :)

Yeah, I know, that's how I've been testing it. Didn't really consider it as a bug though, as you can just as easily select someone else who's not really their mother as well. So you either need some kind of complicated mother-ID checking mechanism or need to just rely on the data entry being accurate. Doesn't seem worth adding protection against one particular special case when there's so many other ways to input inaccurate info.

**Comment by clemens-msupply** [comment]
> > Found one bug: you can select yourself as mother :)
> 
> Yeah, I know, that's how I've been testing it. Didn't really consider it as a bug though, as you can just as easily select someone else who's not really their mother as well. So you either need some kind of complicated mother-ID checking mechanism or need to just rely on the data entry being accurate. Doesn't seem worth adding protection against one particular special case when there's so many other ways to input inaccurate info.

fair enough, on the other hand its an easy fix for this special case...

**Review by clemens-msupply — APPROVED** [review]

**Comment by CarlosNZ** [comment]
> on the other hand its an easy fix for this special case...

Except we don't know that this would be wanted in all cases. This is supposed to be a generic "patientByCode" lookup, not necessarily a "Mother-selector", and there's plenty of possibe use cases where you might allow looking up the current patient. In which case it would need to be a configurable option (`allowCurrentPatient`), but that seems overkill for now.

**Comment by mark-prins** [comment]
fun fact - you can select a patient to be their own mother:

Screenshot 2023-02-28 at 3 07 06 PM

**Comment by mark-prins** [comment]
> Found one bug: you can select yourself as mother :)

I was about to comment with the same thing!
Agree - that a patient lookup control which allows self-selection, is a bit odd.

File: client/packages/programs/src/JsonForms/components/Search/SearchWithUserSource.tsx
Comment (mark-prins) [inline client/packages/programs/src/JsonForms/components/Search/SearchWithUserSource.tsx:139]: even when laid out nicely like this, I think that nested ternary operators are evil. It's too hard to understand what the intention is.

```suggestion

  const getNoOptionsText = () => {
    switch (true) {
      case loading:
        return t('control.search.searching-label');
      case searchText.length < MIN_CHARS:
        return t('control.search.below-min-chars', { minChars: MIN_CHARS });
      default:
        return noResultsText;
    }
  };
  ...
              noOptionsText={getNoOptionsText()}
```

**Review by mark-prins — APPROVED** [review]

**Comment by mark-prins** [comment]
The min characters message is much nicer to use - very clear whats going on. Changes look good thanks!
I'd prefer to remove the nested ternary operator though 😉 
And remove self-selection for patients.. but 🤷

File: client/packages/common/src/utils/formatters/JSXFormatters.tsx
Comment (CarlosNZ) [reply client/packages/common/src/utils/formatters/JSXFormatters.tsx]: I've added a "removeEmptyLines" regex utility in the subsequent PR to achieve this.

**Comment by CarlosNZ** [comment]
> I'd prefer to remove the nested ternary operator though 😉

Done: 64a58c5f5

> And remove self-selection for patients.. but 🤷

See my [comment](https://github.com/openmsupply/open-msupply/pull/1223#issuecomment-1447377427) above.  I remain unconvinced that excluding the current user or the current patient should be the default behaviour for a general "patientByCode" lookup.

I think this is ready to go now, but I'll wait until #1230 is fully approved and merge them both together.

**Comment by mark-prins** [comment]
> I'll wait until https://github.com/openmsupply/open-msupply/pull/1230 is fully approved and merge them both together.

I was waiting until this one was merged and the other branch updated with these changes before looking at that PR!
That then gives us a catch-22.. looks like (2) has been approved already, so I'll just have a look over once I've reviewed ARV. don't wait for me tho

**Pull request merged**
